src/IO/parseDateTimeBestEffort.cpp

```diff
@@ -585,8 +585,7 @@ ReturnType parseDateTimeBestEffortImpl(
     if (!year)
     {
         time_t now = time(nullptr);
-        UInt16 curr_year = local_time_zone.toYear(now);
-        year = now < local_time_zone.makeDateTime(curr_year, month, day_of_month, hour, minute, second) ? curr_year - 1 : curr_year;
+        year = local_time_zone.toYear(now);
```

Comment:
I guess this was done to workaround parsing last year dates in January, where if you see "Dec 30 ..." you understand they are from the previous year and not the current one.

WDYT about keeping the same logic as before but only comparing the date? If it's the current day of the year or a previous one, it's the current year. If it's a date from a date in the year that hasn't happened yet, then it's from last year.

I don't have strong opinions either way.

Reply:
OTOH, if you get "Feb 29" in any of the 2, it's pretty much obvious which year of the 2 might be (only one at most can be correct)

Reply:
Yes, I am not sure what is better, I already asked @alexey-milovidov about his opinion.
And also wanted to check if we have test for this specific logic.

Reply:
I am also not sure if this is not bug with DateLUT, but if we just want to keep old logic and be safe we can handle 29 February with special logic and choose only one valid year.

Reply:
The approach I'm considering (testing) is to pick the year closest to today, so it might be one of the 3: [current_year -1, current_year, current_year + 1

Reply:
I think we should not change default behaviour, and introduce backward-incompatible changes just to fix test.
I think we need either to fix DateLUT implementation or add only logic to handle February 29 in leap year.

Reply:
What do you mean by `fix DateLUT implementation`?

The test is currently fixed (as it's removed). The current logic, which was introduced without any announcement (it used to always be 2000 when missing) is extremely unexpected. It's just that I hadn't noticed it until the test failed.

Reply:
DateLUT `local_time_zone.makeDateTime(curr_year, month, day_of_month, hour, minute, second)` if run for current time must always return the same UNIX timestamp time as `time(nullptr)`, because we rely on this in a lot of other places.
It can be expected that DateLUT does not properly handle leap years, and we need to introduce some logic on top of it, or it is unexpected and we need to fix DateLUT.

Reply:
> It can be expected that DateLUT does not properly handle leap years, and we need to introduce some logic on top of it, or it is unexpected and we need to fix DateLUT.

Those seem like multiple different problems to me:
* If you put '2023-02-29' you get '2023-03-01' because there isn't any '2023-02-29'. We could throw, but that's another breaking change.
* `parseDateTimeeBestEffort` shouldn't set invalid years. Given 'Feb 29th', guessing 2023 is just wrong.
* The current logic to set the year is really unexpected. If I say 'Mar 7th' it's way more likely that I'm talking about next week, and not last year. Maybe we could have some setting to tweak this behaviour (something like `last` for last happened date, `closest` to get the closest date).